### PR TITLE
SONARJAVA-2007 Add exception thrown explicitly in exceptional yields

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 1]
```diff
@@ -43,6 +43,8 @@
   int resultIndex;
   @Nullable
   Constraint resultConstraint;
+  @Nullable
+  String exceptionType;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: If I'm correct this information is sufficient to remove the boolean exception and transform its usage to a method call checking for nullness of this String ?

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 2]
```diff
@@ -161,9 +169,18 @@ public boolean equals(Object obj) {
       || varArgs != other.varArgs) {
       return false;
     }
+    boolean sameExceptionType;
+    if (exceptionType != null) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: see https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#equals(java.lang.Object,%20java.lang.Object)

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 3]
```diff
@@ -161,9 +169,18 @@ public boolean equals(Object obj) {
...
+      sameExceptionType = exceptionType.equals(other.exceptionType);
+    } else {
+      sameExceptionType = other.exceptionType == null;
+    }
+    boolean sameResultConstraint;
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: see https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#equals(java.lang.Object,%20java.lang.Object)

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java (hunk 1 above)
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: As I'm not handling implicit exceptions yet, then the answer is no.
I still need a way to differentiate an exceptional path which isn't aware of the exception causing it, from a happy path.

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 4]
```diff
@@ -155,15 +164,11 @@ public boolean equals(Object obj) {
...
-      || exception != other.exception
-      || resultIndex != other.resultIndex
-      || varArgs != other.varArgs) {
-      return false;
-    }
-    if (resultConstraint != null) {
-      return resultConstraint.equals(other.resultConstraint);
-    }
-    return other.resultConstraint == null;
+    return Arrays.equals(parametersConstraints, other.parametersConstraints)
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: CRITICAL Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**